### ISSUE TITLE
Annotation allowlist-source-range or denylist-source-range in a Service used as backend by multiple paths are applied only to the first one

### ISSUE THREAD

**Issue by kcnpnew** [jcmoraisjr/haproxy-ingress#979]
**Description of the problem**

In version 0.13.4 I noticed what I think is a strange and undocumented behavior when using allowlist-source-range or denylist-source-range annotation in a Service used as backend by multiple paths

Below is a simple test case with two Ingress and one Service

```
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  namespace: test
  name: test1
  annotations:
    kubernetes.io/ingress.class: "haproxy"
spec:
  rules:
  - host: test1.app1
    http:
      paths:
      - path: /test
        pathType: Prefix
        backend:
          service:
            name: test-service
            port:
              name: http
      - path: /test2
        pathType: Prefix
        backend:
          service:
            name: test-service
            port:
              name: http
  - host: test1.app2
    http:
      paths:
      - path: /test
        pathType: Prefix
        backend:
          service:
            name: test-service
            port:
              name: http
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  namespace: test
  name: test2
  annotations:
    kubernetes.io/ingress.class: "haproxy"
spec:
  rules:
  - host: test2.app
    http:
      paths:
      - path: /test
        pathType: Prefix
        backend:
          service:
            name: test-service
            port:
              name: http
---
apiVersion: v1
kind: Service
metadata:
  name: test-service
  labels:
    app: test
  namespace: test
  annotations:
    ingress.kubernetes.io/allowlist-source-range: "10.1.0.0/16"
spec:
  selector:
    app: test
  ports:
  - name: http
    port: 80
    targetPort: 8080

```

**Expected behavior**

As the annotation was applied in the Service I was hoping all paths using that Service as backend would get the source restriction rules.

**Actual behavior**

The source restriction rules are limited to the first path using the Service as a backend.

The generated backend section at the haproxy config is.

```
backend test_test-service_8080
    mode http
    balance roundrobin
    acl https-request ssl_fc
    # path01 = test1.app1/test
    # path02 = test1.app1/test2
    # path03 = test1.app2/test
    # path04 = test2.app/test
    http-request set-var(txn.pathID) var(req.base),map_dir(/etc/haproxy/maps/_back_test_test-service_8080_idpath__prefix.map)
    acl allow_rule_src0 src 10.1.0.0/16
    http-request deny if { var(txn.pathID) path01 } !allow_rule_src0
    http-response set-header Strict-Transport-Security "max-age=15768000; includeSubDomains; preload" if https-request
    server srv001 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv006 127.0.0.1:1023 disabled weight 1 check inter 2s
```

Even though allowlist-source-range has a Path scope, applying it in a Service should work to all paths using it as backend, not only the first one, because the Service has no knowledge in which paths it will be applied. 

This behavior brings another strange effect in this test case.

If we delete the first ingress

```
kubectl -n test delete ingress test1
```

the source restriction rule migrates to next first path available (In this case in the second ingress). But then only one path using the Service remains (the only one in the second ingress) and the rule generated is simpler

```
backend test_test-service_8080
    mode http
    balance roundrobin
    acl https-request ssl_fc
    acl allow_rule_src0 src 10.1.0.0/16
    http-request deny if !allow_rule_src0
    http-response set-header Strict-Transport-Security "max-age=15768000; includeSubDomains; preload" if https-request
    server srv001 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv002 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv003 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv004 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv005 127.0.0.1:1023 disabled weight 1 check inter 2s
    server srv006 127.0.0.1:1023 disabled weight 1 check inter 2s

```

So a change in a unrelated object (delete first ingress) can toggle a restriction in another object (first path in the second ingress)

The same behavior happens with denylist-source-range

**Comment by jcmoraisjr** [comment]
Thanks for reporting! I can confirm this behavior on all supported branches with all path scoped keys. This is happening because service annotations are added when the service is referenced for the fist time. The key should be added as non scoped, which is not supported in the current implementation, so only the scope which found the service is added. It seems I'll have some fun designing a fix to that.

**Comment by jcmoraisjr** [comment]
This should be fixed on v0.13.10 and v0.14.1. Closing. Please update this same issue if you still have any problem.
